plugins/single_plugins/fisheye.cpp

```diff
@@ -0,0 +1,201 @@
+#include <plugin.hpp>
+#include <output.hpp>
+#include <opengl.hpp>
+#include <debug.hpp>
+#include <animation.hpp>
+#include <render-manager.hpp>
+
+static const char* vertex_shader =
+R"(
+#version 100
+
+attribute mediump vec2 position;
+
+void main() {
+
+    gl_Position = vec4(position.xy, 0.0, 1.0);
+}
+)";
+
+static const char* fragment_shader =
+R"(
+#version 100
+precision mediump float;
+
+uniform vec2 u_resolution;
+uniform vec2 u_mouse;
+uniform float u_radius;
+uniform float u_zoom;
+uniform sampler2D u_texture;
+
+const float PI = 3.1415926535;
+
+void main()
+{
+        float radius = u_radius;
+
+        float zoom = u_zoom;
+        float pw = 1.0 / u_resolution.x;
+        float ph = 1.0 / u_resolution.y;
+
+        vec4 p0 = vec4(u_mouse.x, u_resolution.y - u_mouse.y, 1.0 / radius, 0.0);
+        vec4 p1 = vec4(pw, ph, PI / radius, (zoom - 1.0) * zoom);
+        vec4 p2 = vec4(0, 0, -PI / 2.0, 0.0);
+
+        vec4 t0, t1, t2, t3;
+
+        vec3 tc = vec3(1.0, 0.0, 0.0);
+        vec2 uv = vec2(gl_FragCoord.x, gl_FragCoord.y);
+
+        t1 = p0.xyww - vec4(uv, 0.0, 0.0);
+        t2.x = t2.y = t2.z = t2.w = 1.0 / sqrt(dot(t1.xyz, t1.xyz));
+        t0 = t2 - p0;
+
+        t3.x = t3.y = t3.z = t3.w = 1.0 / t2.x;
+        t3 = t3 * p1.z + p2.z;
+        t3.x = t3.y = t3.z = t3.w = cos(t3.x);
+
+        t3 = t3 * p1.w;
+
+        t1 = t2 * t1;
+        t1 = t1 * t3 + vec4(uv, 0.0, 0.0);
+
+        if (t0.z < 0.0) {
+                t1.x = uv.x;
+                t1.y = uv.y;
+        }
+
+        t1 = t1 * p1 + p2;
+
+        tc = texture2D(u_texture, t1.xy).rgb;
+
+        gl_FragColor = vec4(tc, 1.0);
+}
+)";
+
+class wayfire_invert_screen : public wayfire_plugin_t
+{
+
+    post_hook_t hook;
+    key_callback toggle_cb;
+    wf_duration duration;
+    float radius, target_zoom, current_zoom;
```

Comment:
why do you need current zoom? I think you can just use `duration.progress()` every time you need it, or make it a local variable for `render()`. Same for radius, just use `radius_opt->as_cached_double()` each time you need it, and in this case, you can rename it to just `radius`

Reply:
You're basically right, I just made current_zoom local to render() and got rid of radius_opt.